libs/kotlinlib/src/mill/kotlinlib/ksp/KspModule.scala

```diff
@@ -360,6 +390,108 @@ trait KspModule extends KotlinModule { outer =>
 
   }
 
+  /**
+   * The classloader with [[ksp2InProgramToolsClasspath]], which includes the KSP 2 Worker
+   * and the user defined symbol processors via [[kotlinSymbolProcessors]]. This classloader is a
+   * parent of [[kotlinSymbolProcessorClassloader]] as classes are shared between the [[ksp2ToolsDeps]]
+   * and [[kotlinSymbolProcessors]] (symbol processors depend on the KSP API).
+   */
+  def ksp2WorkerClassloader: Worker[MillURLClassLoader] = Task.Worker {
+    Jvm.createClassLoader(
+      classPath = ksp2InProgramToolsClasspath().map(_.path),
+      parent = getClass.getClassLoader
+    )
+  }
+
+  /**
+   * The in-process worker instance of the KSP 2 processor.
+   * The classloader contains [[ksp2InProgramToolsClasspath]], which includes the KSP 2 Worker
+   * and the user defined symbol processors via [[kotlinSymbolProcessors]]. The classloader used here
+   * is [[ksp2WorkerClassloader]]
+   */
+  def ksp2Worker: Worker[KspWorker] = Task.Worker {
+    ksp2WorkerClassloader()
+      .loadClass("mill.kotlinlib.ksp2.worker.KspWorkerImpl").getConstructor().newInstance()
+      .asInstanceOf[KspWorker]
+  }
+
+  /**
+   * The classloader with [[kotlinSymbolProcessorsResolved]] classpath. This is a child of
+   * [[ksp2Worker]] classloader, as classes are shared between the [[ksp2ToolsDeps]]  and
+   * [[kotlinSymbolProcessors]].
+   *
+   * For more info see reference implementation: [[https://github.com/google/ksp/blob/main/docs/ksp2entrypoints.md]]
+   */
+  def kotlinSymbolProcessorClassloader: Worker[MillURLClassLoader] = Task.Worker {
+    Jvm.createClassLoader(
+      kotlinSymbolProcessorsResolved().map(_.path),
+      parent = ksp2Worker().getClass.getClassLoader
+    )
+  }
```

Comment:
I still don't get, why the tool itself needs to be part of this classpath. Isn't this classpath fed to the tool? So, why should I feed the path of the tool to the tool itself? This is rather unusual, and also doesn't match the logic used for KSP 1 and KSP 2 CLI.

Reply:
The tool, similarly to the main approach, has the KSP classpath, this are symbol-processor-aa deps and co (kspToolDeps). This allows it to reference the KSP APIs without knowing the actual user defined processors (e.g. dagger-compiler). 
The discovery starts from the ServiceLoader loading SymbolProcessorProvider (s) . This will discover any KSP providers from the processor classpath (the kspToolDeps). Then the KSP API, accepts a list of symbol processor providers, if the classloaders are independent (no parent relationship) the cast fails.

Is there a different/better way of doing it?

Reply:
Got it. (I somehow don't feel comfortable with the `ksp2Worker().getClass.getClassloader`, although I can't tell right now, why.) Let's see how it works out.

Reply:
I know what you mean, I think it's equivalent with the main, because I was wondering how I was getting class cast errors but the KSPJvmMain with

```kotlin
    val processorClassloader = URLClassLoader(classpath.map { File(it).toURI().toURL() }.toTypedArray())
```

didn't, and it seems that this constructor resolves in:

```java
    protected ClassLoader() {
        this(checkCreateClassLoader(), null, getSystemClassLoader());
    }
```

which is the jvm classpath when it starts, so from then on , the 2 classloaders are basically the same in both cases

Reply:
I see. Lets use `ksp2WorkerClassloader()` as parent them, which is a bit more direct.